File: leetcode/python/823_binary_trees_with_factors.py
from math import sqrt


class Solution:
    def numFactoredBinaryTrees(self, arr: list[int]) -> int:
        arr.sort()
        n = len(arr)
        dp = [1] * n
        index = {arr[i]: i for i in range(n)}
        for i in arr[1:]:
            upper = sqrt(i)
            for j in arr[:i]:
                if j > upper:
                    break

                k, rem = divmod(i, j)
                if rem == 0 and k in index:
                    perm = dp[index[j]] * dp[index[k]]
                    dp[index[i]] += 2 * perm if j != k else perm
        return sum(dp) % (10 ** 9 + 7)

s = Solution()
print(s.numFactoredBinaryTrees([2, 4]))
print(s.numFactoredBinaryTrees([2, 4, 5, 10]))
print(s.numFactoredBinaryTrees([18,3,6,2]))

File: leetcode/typescript/823_binary_trees_with_factors.ts
function numFactoredBinaryTrees(arr: number[]): number {
    arr.sort((a, b) => a - b);
    const index = new Map<number, number>();
    for (let i = 0; i < arr.length; i++) {
        index.set(arr[i], i);
    }
    const dp = new Array(arr.length).fill(1);
    for (const [i, n] of arr.entries()) {
        const upper = Math.sqrt(n);
        for (let l = 0; arr[l] <= upper; l++) {
            const rem = n % arr[l];
            const r = index.get(n / arr[l]);
            if (rem === 0 && r !== undefined) {
                dp[i] += dp[l] * dp[r];
                if (arr[r] !== arr[l]) {
                    dp[i] += dp[l] * dp[r];
                }
                dp[i] %= (10 ** 9 + 7);
            }
            
        }
    }

    return dp.reduce((a, b) => (a + b) % (10 ** 9 + 7), 0);
};